Analytics 4 setup component

zutigrm opened the pull request:
## Summary

<!-- Please reference the issue this PR addresses in the following list. -->
Addresses issue:

- #7929

## Relevant technical choices

* Additionally migrated `AccountCreateLegacy` component as part of this issue, as agreed with Nahid

## PR Author Checklist

- [x] My code is tested and passes existing unit tests.
- [ ] My code has an appropriate set of unit tests which all pass.
- [ ] My code is backward-compatible with WordPress 5.2 and PHP 5.6.
- [ ] My code follows the [WordPress](https://make.wordpress.org/core/handbook/best-practices/coding-standards/) coding standards.
- [ ] My code has proper inline documentation.
- [x] I have added a QA Brief on the issue linked above.
- [x] I have signed the Contributor License Agreement (see <https://cla.developers.google.com/>).

---------------

_Do not alter or remove anything below. The following sections will be managed by moderators only._

## Code Reviewer Checklist

- [x] Run the code.
- [x] Ensure the acceptance criteria are satisfied.
- [x] Reassess the implementation with the IB.
- [x] Ensure no unrelated changes are included.
- [x] Ensure CI checks pass.
- [x] Check Storybook where applicable.
- [x] Ensure there is a QA Brief.

## Merge Reviewer Checklist

- [x] Ensure the PR has the correct target branch.
- [x] Double-check that the PR is okay to be merged.
- [x] Ensure the corresponding issue has a ZenHub release assigned.
- [x] Add a changelog message to the issue.

nfmohit commented on assets/js/modules/analytics-4/components/setup/SetupForm.js:
```diff
@@ -0,0 +1,166 @@
+
+	const { setValues } = useDispatch( CORE_FORMS );
+	const { submitChanges } = useDispatch( MODULES_ANALYTICS_4 );
+	// @TODO remove when analytics 4 is completely decoupled from analytics module.
+	const { submitChanges: submitChangesLegacy } =
+		useDispatch( MODULES_ANALYTICS );
```
This is not necessary as this component will not be surfaced for the end user until the modules are completely decoupled (as explained above).

```suggestion
```

nfmohit commented on assets/js/modules/analytics-4/components/setup/SetupForm.js:
```diff
@@ -0,0 +1,166 @@
+
+			const { error } = await submitChanges();
+
+			// @TODO remove when analytics 4 is completely decoupled from analytics module.
+			if ( ! error ) {
+				await submitChangesLegacy();
+			}
```
```suggestion
```

nfmohit commented on assets/js/modules/analytics-4/components/setup/SetupForm.js:
```diff
@@ -0,0 +1,166 @@
+	ENHANCED_MEASUREMENT_FORM,
+	MODULES_ANALYTICS_4,
+} from '../../datastore/constants';
+import { MODULES_ANALYTICS } from '../../../analytics/datastore/constants';
```
```suggestion
```

nfmohit commented on assets/js/modules/analytics-4/components/common/AccountCreateLegacy.js:
```diff
@@ -0,0 +1,133 @@
+	);
+
+	const { resetAccountSummaries } = useDispatch( MODULES_ANALYTICS_4 );
+	const { resetAccounts: resetUAAccounts } = useDispatch( MODULES_ANALYTICS );
```
Not needed as this will not be used until the modules are decoupled in #7932.

```suggestion
```

nfmohit commented on assets/js/modules/analytics-4/components/common/AccountCreateLegacy.js:
```diff
@@ -0,0 +1,133 @@
+
+	const { resetAccountSummaries } = useDispatch( MODULES_ANALYTICS_4 );
+	const { resetAccounts: resetUAAccounts } = useDispatch( MODULES_ANALYTICS );
+	const refetchAccountsHandler = useCallback( () => {
+		resetAccountSummaries();
+		// Backward compatibility until Anlytics 4 is decoupled fully from Analytics module
+		// as accountID is still in Analytics datastore, and it is not reset with Analytics 4
+		// `resetAccountSummaries` action. So when re-fething the accounts `accountID` will remain
+		// `account_create` and hence not render the `SetupForm` component.
+		resetUAAccounts();
+	}, [ resetAccountSummaries, resetUAAccounts ] );
```
```suggestion
	const refetchAccountsHandler = useCallback( () => {
		resetAccountSummaries();
	}, [ resetAccountSummaries ] );
```

nfmohit commented on assets/js/modules/analytics-4/components/common/AccountCreateLegacy.js:
```diff
@@ -0,0 +1,133 @@
+import { Button, ProgressBar } from 'googlesitekit-components';
+import { trackEvent } from '../../../../util';
+import { MODULES_ANALYTICS_4, ACCOUNT_CREATE } from '../../datastore/constants';
+import { MODULES_ANALYTICS } from '../../../analytics/datastore/constants';
```
```suggestion
```

nfmohit commented on assets/js/modules/analytics-4/datastore/tagmanager.js:
```diff
@@ -0,0 +1,75 @@
+	),
+};
+
+const store = Data.combineStores( {
+	selectors: baseSelectors,
+} );
+
+export const initialState = store.initialState;
+export const actions = store.actions;
+export const controls = store.controls;
+export const reducer = store.reducer;
+export const resolvers = store.resolvers;
+export const selectors = store.selectors;
+
+export default store;
```
As this partial contains just a selector, we don't have to use `Data.combineStores`.

```suggestion
const store = {
	selectors: baseSelectors,
};

export default store;
```

nfmohit commented on assets/js/modules/analytics-4/components/setup/SetupForm.stories.js:
```diff
@@ -0,0 +1,108 @@
+				registry
+					.dispatch( MODULES_ANALYTICS_4 )
+					.selectAccount( accountID );
+
+				// @TODO Remove after analytics 4 is decoupled completely from analytics module.
+				registry
+					.dispatch( MODULES_ANALYTICS )
+					.receiveGetProperties( [], { accountID } );
+				registry
+					.dispatch( MODULES_ANALYTICS )
+					.selectAccount( accountID );
```
```suggestion

```

nfmohit reviewed (changes requested):
Thank you for the hard work here, @zutigrm.

I have left a detailed CR feedback. The main theme of the feedback is that we don't want to surface this setup component for the end user just yet. It should just be a shadow of the `analytics` setup component and we want to surface this as part of #7932 when we flip the switch. I apologise, this should've probably been clearer in the ACs. This should be similar to how #7931 was executed.

Could you address the CR feedback and update the tests & stories accordingly?

Besides, could you also update the QAB accordingly and extend the QA:Eng brief so that an engineer can replace the `SetupMain` component in `analytics` module registration to test the new component and ensure it looks and works well?

Thank you!

zutigrm commented on assets/js/modules/analytics/components/setup/SetupMain.js:
```diff
@@ -33,13 +33,13 @@ import { useEffect, useState } from '@wordpress/element';
 import useExistingTagEffectGA4 from '../../../analytics-4/hooks/useExistingTagEffect';
 import { MODULES_ANALYTICS_4 } from '../../../analytics-4/datastore/constants';
 import { AccountCreate, AccountCreateLegacy } from '../common';
+import { SetupForm } from '../../../analytics-4/components/setup';
```
Reverted to the `analytics` module `SetupForm`

zutigrm commented:
Thanks @nfmohit I updated the PR.

Note that some js tests have to include some data to `analytics` datastore in order to work, so I can't remove them, so I added `@TODO` comment so it can be removed in the last issue that decouples the analytics 4 module

nfmohit commented on assets/js/modules/analytics-4/components/setup/SetupForm.test.js:
```diff
@@ -0,0 +1,459 @@
+		const createPropertyRegexp = new RegExp(
+			'/analytics-4/data/create-property'
+		);
+
+		fetchMock.post( createPropertyRegexp, {
+			status: 403,
+			body: {
+				code: 403,
+				error: 'Insufficient permissions',
+			},
+		} );
+
+		const finishSetup = jest.fn();
+		const { getByRole, waitForRegistry } = render(
+			<SetupForm finishSetup={ finishSetup } />,
+			{
+				registry,
+			}
+		);
+		await waitForRegistry();
+
+		// Ensure the form rendered successfully.
+		expect(
+			getByRole( 'button', { name: /Configure Analytics/i } )
+		).toBeInTheDocument();
+
+		// While not strictly needed, add waits to match the successful auto-submit test case to help avoid a false positive result.
+		await waitForRegistry();
+		await act( async () => {
+			await waitForDefaultTimeouts();
+		} );
+
+		// Create property should have only been called once.
+		expect( fetchMock ).toHaveFetchedTimes( 1, createPropertyRegexp );
+		// Setup was not successful, so the finish function should not be called.
+		expect( finishSetup ).not.toHaveBeenCalled();
+		// Expect a console error due to the API error (otherwise this test will fail).
+		expect( console ).toHaveErrored();
+	} );
+} );
```
In order to minimize references to the legacy `analytics` datastore, may I suggest changing the test cases according to the following?

```suggestion
describe( 'SetupForm', () => {
	let registry;

	beforeEach( () => {
		registry = createTestRegistry();
		provideSiteInfo( registry );
		provideUserAuthentication( registry );
		provideModules( registry, [ { slug: 'analytics-4', active: true } ] );
	} );

	it( 'renders the form with a progress bar when GTM containers are not resolved', () => {
		provideModules( registry, [
			{ slug: 'analytics-4', active: true, connected: true },
			{ slug: 'tagmanager', active: true, connected: true },
		] );
		registry.dispatch( MODULES_ANALYTICS_4 ).setSettings( {} );
		registry.dispatch( MODULES_TAGMANAGER ).setSettings( {} );

		const { container, getByRole } = render( <SetupForm />, { registry } );

		expect( container ).toMatchSnapshot();

		expect( getByRole( 'progressbar' ) ).toBeInTheDocument();
	} );

	it( 'renders the form correctly', async () => {
		registry.dispatch( MODULES_ANALYTICS_4 ).setSettings( {} );
		registry.dispatch( MODULES_TAGMANAGER ).setSettings( {} );
		registry
			.dispatch( MODULES_ANALYTICS_4 )
			.receiveGetAccountSummaries( accountSummaries );
		registry
			.dispatch( MODULES_ANALYTICS_4 )
			.receiveGetProperty( properties[ 0 ], {
				propertyID,
			} );
		registry
			.dispatch( MODULES_ANALYTICS_4 )
			.receiveGetWebDataStreamsBatch( webDataStreamsBatch, {
				propertyIDs: [ propertyID ],
			} );
		registry
			.dispatch( MODULES_ANALYTICS_4 )
			.receiveGetEnhancedMeasurementSettings(
				{
					...defaultEnhancedMeasurementSettings,
					streamEnabled: false,
				},
				{
					propertyID,
					webDataStreamID,
				}
			);

		// @TODO: This is temporarily needed here as the AccountSelect component
		// still relies on the `analytics` datastore. This should be removed
		// once the two Analytics modules are entirely decoupled.
		registry.dispatch( MODULES_ANALYTICS ).setAccountID( accountID );

		const { container, getByText, waitForRegistry } = render(
			<SetupForm />,
			{
				registry,
			}
		);
		await waitForRegistry();

		expect( container ).toMatchSnapshot();

		expect( getByText( 'Account' ) ).toBeInTheDocument();
		expect( getByText( 'Property' ) ).toBeInTheDocument();
		expect( getByText( 'Web Data Stream' ) ).toBeInTheDocument();
	} );

	it( 'submits the form upon pressing the CTA', async () => {
		registry.dispatch( MODULES_ANALYTICS_4 ).setSettings( {} );
		registry.dispatch( MODULES_TAGMANAGER ).setSettings( {} );
		registry
			.dispatch( MODULES_ANALYTICS_4 )
			.receiveGetAccountSummaries( accountSummaries );
		registry
			.dispatch( MODULES_ANALYTICS_4 )
			.receiveGetProperty( properties[ 0 ], {
				propertyID,
			} );
		registry
			.dispatch( MODULES_ANALYTICS_4 )
			.receiveGetWebDataStreamsBatch( webDataStreamsBatch, {
				propertyIDs: [ propertyID ],
			} );
		registry
			.dispatch( MODULES_ANALYTICS_4 )
			.receiveGetEnhancedMeasurementSettings(
				defaultEnhancedMeasurementSettings,
				{
					propertyID,
					webDataStreamID,
				}
			);

		await registry
			.dispatch( MODULES_ANALYTICS_4 )
			.selectAccount( accountID );

		// @TODO: This is temporarily needed here as the AccountSelect component
		// still relies on the `analytics` datastore. This should be removed
		// once the two Analytics modules are entirely decoupled.
		registry.dispatch( MODULES_ANALYTICS ).setAccountID( accountID );

		const finishSetup = jest.fn();
		const { getByRole, waitForRegistry } = render(
			<SetupForm finishSetup={ finishSetup } />,
			{
				registry,
			}
		);
		await waitForRegistry();

		act( () => {
			// It doesn't seem possible to show the dropdown menu within the test by clicking on the dropdown in the usual way.
			// Therefore, we simulate the click on the dropdown menu item directly, despite it being hidden.
			fireEvent.click(
				getByRole( 'menuitem', {
					name: /Example Com/i,
					hidden: true,
				} )
			);
		} );

		registry
			.dispatch( CORE_USER )
			.receiveGetDismissedTours( [ ga4ReportingTour.slug ] );

		const updateAnalyticsSettingsRegexp = new RegExp(
			'/analytics-4/data/settings'
		);

		const dismissItemEndpointRegexp = new RegExp(
			'^/google-site-kit/v1/core/user/data/dismiss-item'
		);

		muteFetch( updateAnalyticsSettingsRegexp );

		fetchMock.post( dismissItemEndpointRegexp, {
			status: 200,
			body: JSON.stringify( [
				ENHANCED_MEASUREMENT_ACTIVATION_BANNER_DISMISSED_ITEM_KEY,
			] ),
		} );

		act( () => {
			fireEvent.click(
				getByRole( 'button', { name: /Configure Analytics/i } )
			);
		} );

		await waitForRegistry();

		// An additional wait is required in order for all resolvers to finish.
		await act( waitForDefaultTimeouts );

		expect( fetchMock ).toHaveFetchedTimes(
			1,
			updateAnalyticsSettingsRegexp
		);

		expect( finishSetup ).toHaveBeenCalledTimes( 1 );
	} );

	it( 'auto-submits the form', async () => {
		registry.dispatch( MODULES_ANALYTICS_4 ).setSettings( {} );
		registry.dispatch( MODULES_TAGMANAGER ).setSettings( {} );
		registry.dispatch( MODULES_ANALYTICS_4 ).receiveGetExistingTag( null );
		registry.dispatch( MODULES_ANALYTICS_4 ).receiveGetAccountSummaries(
			accountSummaries.map( ( account ) => ( {
				...account,
				propertySummaries: [],
			} ) )
		);
		registry
			.dispatch( MODULES_ANALYTICS_4 )
			.receiveGetProperty( properties[ 0 ], {
				propertyID,
			} );
		registry
			.dispatch( MODULES_ANALYTICS_4 )
			.receiveGetWebDataStreamsBatch( webDataStreamsBatch, {
				propertyIDs: [ propertyID ],
			} );
		registry
			.dispatch( MODULES_ANALYTICS_4 )
			.receiveGetEnhancedMeasurementSettings(
				defaultEnhancedMeasurementSettings,
				{
					propertyID,
					webDataStreamID,
				}
			);

		await registry
			.dispatch( MODULES_ANALYTICS_4 )
			.selectAccount( accountID );

		// @TODO: This is temporarily needed here as the AccountSelect component
		// still relies on the `analytics` datastore. This should be removed
		// once the two Analytics modules are entirely decoupled.
		registry.dispatch( MODULES_ANALYTICS ).setAccountID( accountID );

		// Simulate an auto-submit case where the user is returning to the page
		// after granting extra scopes necessary to submit.
		// In this situation, the autoSubmit is set before the user goes to oAuth,
		// store state is snapshotted, and then restored upon returning.
		registry
			.dispatch( CORE_FORMS )
			.setValues( FORM_SETUP, { autoSubmit: true } );

		provideUserAuthentication( registry, {
			grantedScopes: [ EDIT_SCOPE ],
		} );

		registry
			.dispatch( CORE_USER )
			.receiveGetDismissedTours( [ ga4ReportingTour.slug ] );

		const createPropertyRegexp = new RegExp(
			'/analytics-4/data/create-property'
		);

		const createWebDataStreamRegexp = new RegExp(
			'/analytics-4/data/create-webdatastream'
		);

		const updateAnalyticsSettingsRegexp = new RegExp(
			'/analytics-4/data/settings'
		);

		const accountSummariesRegexp = new RegExp(
			'/analytics-4/data/account-summaries'
		);

		fetchMock.post( createPropertyRegexp, {
			status: 200,
			body: properties[ 0 ],
		} );

		fetchMock.post( createWebDataStreamRegexp, {
			status: 200,
			body: webDataStreams[ 2 ],
		} );

		fetchMock.get( accountSummariesRegexp, {
			status: 200,
			body: accountSummaries,
		} );

		muteFetch( updateAnalyticsSettingsRegexp );

		const finishSetup = jest.fn();
		const { getByRole, waitForRegistry } = render(
			<SetupForm finishSetup={ finishSetup } />,
			{
				registry,
			}
		);
		await waitForRegistry();

		// An additional wait is required in order for all resolvers to finish.
		await act( async () => {
			await waitForDefaultTimeouts();
		} );

		// Ensure the form rendered successfully.
		expect(
			getByRole( 'button', { name: /Configure Analytics/i } )
		).toBeInTheDocument();

		await waitForRegistry();

		// An additional wait is required in order for all resolvers to finish.
		await act( async () => {
			await waitForDefaultTimeouts();
		} );

		expect( fetchMock ).toHaveFetchedTimes( 1, createPropertyRegexp );
		expect( fetchMock ).toHaveFetchedTimes( 1, createWebDataStreamRegexp );
		expect( fetchMock ).toHaveFetchedTimes(
			1,
			updateAnalyticsSettingsRegexp
		);

		expect( finishSetup ).toHaveBeenCalledTimes( 1 );
	} );

	it( 'auto-submits the form only once in the case of an error', async () => {
		registry.dispatch( MODULES_TAGMANAGER ).setSettings( {} );
		registry.dispatch( MODULES_ANALYTICS_4 ).receiveGetExistingTag( null );
		registry.dispatch( MODULES_ANALYTICS_4 ).receiveGetAccountSummaries(
			accountSummaries.map( ( account ) => ( {
				...account,
				propertySummaries: [],
			} ) )
		);
		registry
			.dispatch( MODULES_ANALYTICS_4 )
			.receiveGetWebDataStreamsBatch( webDataStreamsBatch, {
				propertyIDs: [ propertyID ],
			} );

		await registry
			.dispatch( MODULES_ANALYTICS_4 )
			.selectAccount( accountID );

		// @TODO: This is temporarily needed here as the AccountSelect component
		// still relies on the `analytics` datastore. This should be removed
		// once the two Analytics modules are entirely decoupled.
		registry.dispatch( MODULES_ANALYTICS ).setAccountID( accountID );

		// Simulate an auto-submit case where the user is returning to the page
		// after granting extra scopes necessary to submit.
		// In this situation, the autoSubmit is set before the user goes to oAuth,
		// store state is snapshotted, and then restored upon returning.
		registry
			.dispatch( CORE_FORMS )
			.setValues( FORM_SETUP, { autoSubmit: true } );

		provideUserAuthentication( registry, {
			grantedScopes: [ EDIT_SCOPE ],
		} );

		registry
			.dispatch( CORE_USER )
			.receiveGetDismissedTours( [ ga4ReportingTour.slug ] );

		const createPropertyRegexp = new RegExp(
			'/analytics-4/data/create-property'
		);

		fetchMock.post( createPropertyRegexp, {
			status: 403,
			body: {
				code: 403,
				error: 'Insufficient permissions',
			},
		} );

		const finishSetup = jest.fn();
		const { getByRole, waitForRegistry } = render(
			<SetupForm finishSetup={ finishSetup } />,
			{
				registry,
			}
		);
		await waitForRegistry();

		// Ensure the form rendered successfully.
		expect(
			getByRole( 'button', { name: /Configure Analytics/i } )
		).toBeInTheDocument();

		// While not strictly needed, add waits to match the successful auto-submit test case to help avoid a false positive result.
		await waitForRegistry();
		await act( async () => {
			await waitForDefaultTimeouts();
		} );

		// Create property should have only been called once.
		expect( fetchMock ).toHaveFetchedTimes( 1, createPropertyRegexp );
		// Setup was not successful, so the finish function should not be called.
		expect( finishSetup ).not.toHaveBeenCalled();
		// Expect a console error due to the API error (otherwise this test will fail).
		expect( console ).toHaveErrored();
	} );
} );
```

zutigrm replied:
Changing the tests to lessen the `analytics` module involvement would kinda beat the goal of cloning the `analytics-4` components and datastore, and having that 1:1 copy that should work as before after the switch is made. As these tests should be able to work as is after `analytics` is removed from codebase and this test file, along with replaced datastore in few instances. Besides, 3 out of 5 tests even when modified per suggestion above, are still failing due to dependency on `analytics` data, and are still making a request to fetch it.

nfmohit replied:
Thank you for sharing your thoughts.

> Changing the tests to lessen the `analytics` module involvement would kinda beat the goal of cloning the `analytics-4` components and datastore, and having that 1:1 copy that should work as before after the switch is made. As these tests should be able to work as is after `analytics` is removed from codebase and this test file, along with replaced datastore in few instances.

I absolutely understand. It's just that we want to reduce dependency on the legacy `analytics` module as much as possible at this point and make the implementation as close as possible to the final expectation to reduce the burden on #7932.

If you take a look at my suggestion, this is how the test suite will ultimately turn out to be when #7932 is implemented. The suggestion practically removes the fetch requests by the `analytics` `selectAccount` action and its side effects, as they'll ultimately not exist when we no longer use this module.

> Besides, 3 out of 5 tests even when modified per suggestion above, are still failing due to dependency on `analytics` data, and are still making a request to fetch it.

I just ran a test with my suggestions applied, and all of them seem to pass, even without any changes to the snapshots. See screenshot:

![image](https://github.com/google/site-kit-wp/assets/20284937/bf533e59-ac7c-46ec-805a-c828ce7d4b20)

Please let me know what you think, thanks!

CC: @jimmymadon as co-lead, to share thoughts on my suggestions. Thanks!

zutigrm replied:
Makes sense, updating. Originally when I opened the file, didn;t notice it is the version from old `analytics` module 🤦

nfmohit reviewed (changes requested):
Thank you for the brilliant work here, @zutigrm. I have left a few additional comments, after addressing which, we should be good to merge this.

Besides, I'd recommend updating the QA:Eng Brief to update the import for the `SetupMain` component in `assets/js/modules/analytics/index.js` instead of updating the import for `SetupForm` in `assets/js/modules/analytics/components/setup/SetupMain.js`.

Please reach out if you have any questions, thank you!

zutigrm commented:
Thanks @nfmohit . PR and QAB updated

nfmohit commented on assets/js/modules/analytics-4/datastore/service.test.js:
Sounds good. It just struck me because the service datastore partial isn't related to the scope of this issue, and does throw itself as an unrelated change. But it is just tests and a change for good, so I think it's alright.
